### PR TITLE
WIP: ConvertFrom/To-CliXml

### LINKED ISSUE #8311: Allow ConvertTo-Xml to emit CLIXML (the same XML format as Export-Clixml)
# Summary of the new feature/enhancement

Follow-up from https://github.com/PowerShell/PowerShell/issues/8308#issuecomment-439737699, quoting @SteveL-MSFT:

> Ideally, it seems like we should have had `ConvertTo-Xml` take a `-Format` parameter that supported CliXml and other xml formats rather than having this confusion between `ExportTo-CliXml` and `ConvertTo-Xml` having incompatible xml formats.

Note that the documentation is lacking, as of this writing, as it mistakenly suggests that `ConvertTo-Xml` is the in-memory equivalent of `Export-CliXml` - see https://github.com/PowerShell/PowerShell-Docs/issues/3310

_Update_: Adding dedicated `ConvertTo-CliXml` and `ConvertFrom-CliXml` cmdlets has been requested and green-lit (#3898), though the associated PR has gone stale: https://github.com/PowerShell/PowerShell/pull/5815

In light of that, it may be better to leave `ConvertTo-Xml` alone and simply fix / amend its documentation.

### PR DESCRIPTION
**Pull request by IISResetMe**
Fix #8311

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 1]
```diff
@@ -120,6 +120,8 @@ public SwitchParameter NoClobber
```
Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 4 failures in `PowerShell-CI-macos`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```
XmlCommand DRT basic functionality Tests.Import-Clixml -IncludeTotalCount always returns unknown total count
```powershell
Expected strings to be the same, but they were different.
Expected length: 19
Actual length:   0
Strings differ at index 0.
Expected: 'Unknown total count'
But was:  ''
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 120
120: 		$out[0].ToString() | Should -BeExactly "Unknown total count"
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for simple types
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 126
126: 		$out.Count | Should -Be 2
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for collections
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 136
136: 		$out2.Count | Should -Be 2
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 6 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator=<=}, @{InputObject=ConvertTo-Clixml; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 540
540:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator=<=}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 564
564:         $result | Should -BeNullOrEmpty
```
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```
XmlCommand DRT basic functionality Tests.Import-Clixml -IncludeTotalCount always returns unknown total count
```powershell
Expected strings to be the same, but they were different.
Expected length: 19
Actual length:   0
Strings differ at index 0.
Expected: 'Unknown total count'
But was:  ''
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\XMLCommand.Tests.ps1: line 120
120: 		$out[0].ToString() | Should -BeExactly "Unknown total count"
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for simple types
```powershell
Expected 2, but got 4.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\XMLCommand.Tests.ps1: line 126
126: 		$out.Count | Should -Be 2
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 4 failures in `PowerShell-CI-linux`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```
XmlCommand DRT basic functionality Tests.Import-Clixml -IncludeTotalCount always returns unknown total count
```powershell
Expected strings to be the same, but they were different.
Expected length: 19
Actual length:   0
Strings differ at index 0.
Expected: 'Unknown total count'
But was:  ''
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 120
120: 		$out[0].ToString() | Should -BeExactly "Unknown total count"
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for simple types
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 126
126: 		$out.Count | Should -Be 2
```
XmlCommand DRT basic functionality Tests.Import-Clixml -First and -Skip work together for collections
```powershell
Expected 2, but got 4.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/XMLCommand.Tests.ps1: line 136
136: 		$out2.Count | Should -Be 2
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 3 failures in `PowerShell-CI-windows`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator=<=}, @{InputObject=ConvertTo-Clixml; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 540
540:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator=<=}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator=<=}).
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\DefaultCommands.Tests.ps1: line 564
564:         $result | Should -BeNullOrEmpty
```
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 1 failures in `PowerShell-CI-linux`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 1 failures in `PowerShell-CI-macos`
CliXml test.Import-CliXML.test follow-up for WinBlue: 161470 - Export-CliXml errors in WhatIf scenarios
```powershell
Object reference not set to an instance of an object.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Utility/clixml.tests.ps1: line 169
169:             Export-Clixml -Path $testPath -InputObject "string" -WhatIf

```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 2 failures in `PowerShell-CI-macos`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 2 failures in `PowerShell-CI-linux`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 2 failures in `PowerShell-CI-macos`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 2 failures in `PowerShell-CI-linux`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 2]
```diff
@@ -362,6 +316,236 @@ protected override void StopProcessing()
...
+        {
+            CreateMemoryStream();
+        }
+
+        /// <summary>
+        /// Override ProcessRecord.
+        /// </summary>
+        protected override void ProcessRecord()
+        {
+            if (_serializer != null)
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: It looks like _serializer is always created in 'CreateMemoryStream', so there is no need to check if it is null.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 3]
```diff
@@ -362,6 +316,236 @@ protected override void StopProcessing()
...
+                _serializer.Serialize(InputObject);
+        }
+
+        /// <summary>
+        /// </summary>
+        protected override void EndProcessing()
+        {
+            if (_serializer != null)
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: Null check appears unnecessary.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 4]
```diff
@@ -623,38 +807,157 @@ private void CleanUp()
...
+        internal XmlWriter _xw;
+
+        internal static XmlWriter CreateXmlWriter(Stream stream)
+        {
+            return XmlWriter.Create(stream, InternalSerializer.GetXmlWriterSettingsForCliXml(ClrFacade.GetDefaultEncoding()));
+        }
+
+        internal void Serialize(object inputObject)
+        {
+            if (_serializer != null)
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: Looks like _serializer is created in the constructor, so don't need null check.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs [hunk 5]
```diff
@@ -679,18 +982,24 @@ private static XmlReader CreateXmlReader(Stream stream)
             return XmlReader.Create(textReader, InternalDeserializer.XmlReaderSettingsForCliXml);
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: This seems wrong as a new reader is created every time the property is accessed.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs (hunk 2 above)
Comment (IISResetMe) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: Yeah, you're right, but it should also be moved completely back to the `ExportXmlHelper` class. 

The idea here is that `ConvertTo/Export-Clixml` and `ConvertFrom/Import-CliXml` become thin wrappers around the helpers which in turn abstracts away the the (de-)serialization interfaces, and the cmdlets themselves just feed a reader or writer to the helper class.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs (hunk 1 above)
Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/XmlCommands.cs]: @IISResetMe, your last commit had 2 failures in `PowerShell-CI-linux`
Verify approved aliases list.All approved Cmdlets present (no new Cmdlets added, no Cmdlets removed)
```powershell
Expected $null or empty, but got @(@{InputObject=ConvertFrom-Clixml; SideIndicator==>}, @{InputObject=ConvertTo-Clixml; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 542
542:         $result | Should -BeNullOrEmpty
```
Verify approved aliases list.All present Cmdlets should have the correct ConfirmImpact
```powershell
Expected $null or empty, but got @(@{Name=Write-Verbose; ConfirmImpact=None; SideIndicator==>}, @{Name=Write-Warning; ConfirmImpact=None; SideIndicator==>}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/DefaultCommands.Tests.ps1: line 566
566:         $result | Should -BeNullOrEmpty
```

**Comment by TravisEz13** [comment]
Please don't mark milestone until merged

**Pull request closed without merging**